Src/FluentAssertions/Equivalency/AssertionResultSet.cs

```diff
@@ -30,25 +30,27 @@ public void AddSet(object key, string[] failures)
         /// </remarks>
         public string[] SelectClosestMatchFor(object key = null)
         {
-            if (!ContainsSuccessfulSet)
+            if (ContainsSuccessfulSet)
             {
-                KeyValuePair<object, string[]> bestMatch = BestResultSets.Any(r => r.Key.Equals(key))
-                    ? BestResultSets.Single(r => r.Key.Equals(key))
-                    : BestResultSets.First();
+                return new string[0];
+            }
+
+            KeyValuePair<object, string[]>[] bestResultSets = BestResultSets();
+
+            KeyValuePair<object, string[]> bestMatch = bestResultSets.FirstOrDefault(r => r.Key.Equals(key));
 
-                return bestMatch.Value;
+            if(bestMatch.Equals(default(KeyValuePair<object, string[]>)))
+            {
+                return bestResultSets[0].Value;
             }
 
-            return new string[0];
+            return bestMatch.Value;
         }
 
-        private KeyValuePair<object, string[]>[] BestResultSets
+        private KeyValuePair<object, string[]>[] BestResultSets()
```

Comment:
If this is turned into a method, it probably should be named `GetBestResultSets` or something alike